# Set Directory for testing purposes
setwd("~/Desktop/CS/Seurat/repos/Seurat.scATAC-seq/src")

# https://satijalab.org/signac/articles/pbmc_vignette.html


# Load Libraries to analyze human data
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Hsapiens.v75)
library(ggplot2)
library(patchwork)
set.seed(1234)

###########################
# PRE-PROCESSING WORKFLOW #
###########################

# Create a Seurat Object using the peak/cell matrix and cell metadata generated by
# `cellranger-atac` and store the path to the fragment file on disk in the Seurat Object
counts <- Seurat::Read10X_h5(filename = "../data/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5")
metadata <- read.csv(file = "../data/atac_v1_pbmc_10k_singlecell.csv", header=TRUE, row.names=1)


chrom_assay <- Signac::CreateChromatinAssay(
                                    counts=counts,
                                    sep=c(":", "-"),
                                    genome="hg19",
                                    fragments="../data/atac_v1_pbmc_10k_fragments.tsv.gz",
                                    min.cells=10,
                                    min.features=200
                                   )

pbmc <- Seurat::CreateSeuratObject(
                          counts=chrom_assay, 
                          assay="peaks", 
                          meta.data=metadata
                          )

pbmc
pbmc[["peaks"]]

Signac::granges(pbmc)

# Extract gene annotations from EnsDb
annotations <- Signac::GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)

# Change to UCSC style since the data was mapped to hg19
GenomeInfoDb::seqlevelsStyle(annotations) <- "UCSC"
GenomeInfoDb::genome(annotations) <- "hg19"

# Add the gene information to the object
Signac::Annotation(pbmc) <- annotations

########################
# COMPUTING QC METRICS #
########################

# Compute nucleosome signal score per cell
pbmc <- Signac::NucleosomeSignal(object=pbmc)

# Compute TSS enrichment score per cell
pbmc <- Signac::TSSEnrichment(object=pbmc, fast=FALSE)

# add blacklist ratio and fraction of reads in peaks
pbmc$pct_reads_in_peaks <- pbmc$peak_region_fragments / pbmc$passed_filters * 100
pbmc$blacklist_ratio <- pbmc$blacklist_region_fragments / pbmc$peak_region_fragments
pbmc$high.tss <- ifelse(pbmc$TSS.enrichment > 2, "High", "Low")

Signac::TSSPlot(pbmc, group.by="high.tss") + NoLegend()


pbmc$nucleosome_group <- ifelse(pbmc$nucleosome_signal > 4, "NS > 4", "NS < 4")
Signac::FragmentHistogram(object=pbmc, group.by="nucleosome_group")

Seurat::VlnPlot(object=pbmc, features=c("pct_reads_in_peaks", "peak_region_fragments", "TSS.enrichment", "blacklist_ratio", "nucleosome_signal"),
               pt.size=0.1,
               ncol=5
               )

pbmc <- subset(x=pbmc, 
               subset=peak_region_fragments > 3000 &
               peak_region_fragments < 20000 &
               pct_reads_in_peaks > 15 &
               blacklist_ratio < 0.05 &
               nucleosome_signal < 4 &
               TSS.enrichment > 2
              )
pbmc

################################################
# NORMALIZATION & LINEAR DIMENSIONAL REDUCTION #
################################################

pbmc <- Signac::RunTFIDF(pbmc)
pbmc <- Signac::FindTopFeatures(pbmc, min.cutoff="q0")
pbmc <- Signac::RunSVD(pbmc)

Signac::DepthCor(pbmc)

#################################################
# NON-LINEAR DIMENSION REDUCTION AND CLUSTERING #
#################################################

pbmc <- Seurat::RunUMAP(object=pbmc, reduction="lsi", dims=2:30)
pbmc <- Seurat::FindNeighbors(object=pbmc, reduction="lsi", dims=2:30)
pbmc <- Seurat::FindClusters(object=pbmc, verbose=FALSE, algorithm=3)
Seurat::DimPlot(object=pbmc, label=TRUE) + Seurat::NoLegend()

#################################
# CREATE A GENE ACTIVITY MATRIX #
#################################

gene.activities <- Signac::GeneActivity(pbmc)

# Add the gene activity matrix to the Seurat object as a new assay and normalize it
pbmc[["RNA"]] <- Seurat::CreateAssayObject(counts=gene.activities)
pbmc <- Seurat::NormalizeData(
                              object = pbmc,
                              assay = "RNA",
                              normalization.method = "LogNormalize",
                              scale.factor = median(pbmc$nCount_RNA)
                             )

Seurat::DefaultAssay(pbmc) <- "RNA"
Seurat::FeaturePlot(
  object = pbmc,
  features = c("MS4A1", "CD3D", "LEF1", "NKG7", "TREM1", "LYZ"),
  pt.size = 0.1,
  max.cutoff = "q95",
  ncol = 3
)

###################################
# INTEGRATING WITH scRNA-seq DATA #
###################################

# Load the pre-processed scRNA-seq data for PBMCs
# Note: 
# Raw Data - https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_10k_v3
# GitHub Code - https://github.com/satijalab/Integration2019/blob/master/preprocessing_scripts/pbmc_10k_v3.R
# RDS Download - https://www.dropbox.com/s/zn6khirjafoyyxl/pbmc_10k_v3.rds?dl=0
pbmc_rna <- readRDS("../data/pbmc_10k_v3.rds")

transfer.anchors <- Seurat::FindTransferAnchors(
  reference = pbmc_rna,
  query = pbmc,
  reduction = "cca"
)

predicted.labels <- Seurat::TransferData(
  anchorset = transfer.anchors,
  refdata = pbmc_rna$celltype,
  weight.reduction = pbmc[["lsi"]],
  dims = 2:30
)

pbmc <- AddMetaData(object=pbmc, metadata=predicted.labels)

plot1 <- Seurat::DimPlot(
  object = pbmc_rna,
  group.by = "celltype",
  label = TRUE,
  repe = TRUE
) + Seurat::NoLegend() + ggplot2::ggtitle("scRNA-seq")

plot2 <- Seurat::DimPlot(
  object = pbmc,
  group.by = "predicted.id",
  label = TRUE,
  repel = TRUE
) + Seurat::NoLegend() + ggtitle("scATAC-seq")

plot1 + plot2

pbmc <- subset(pbmc, idents=14, invert=TRUE)
pbmc <- Seurat::RenameIdents(
  object = pbmc,
  "0" = "CD14 Mono",
  "1" = "CD4 Memory",
  "2" = "CD8 Effector",
  "3" = "CD4 Naive",
  "4" = "CD14 Mono",
  "5" = "DN T",
  "6" = "CD8 Naive",
  "7" = "NK CD56Dim",
  "8" = "pre-B",
  "9" = "CD16 Mono",
  "10" = "pro-B",
  "11" = "DC",
  "12" = "NK CD56bright",
  "13" = "pDC"
)

#########################################################
# FIND DIFFERENTIALLY ACCESSIBLE PEAKS BETWEEN CLUSTERS #
#########################################################

# change back to working with peaks instead of gene activities
Seurat::DefaultAssay(pbmc) <- "peaks"

da_peaks <- Seurat::FindMarkers(
  object = pbmc,
  ident.1 = "CD4 Naive",
  ident.2 = "CD14 Mono",
  min.pct = 0.2,
  test.use = "LR",
  latent.vars = "peak_region_fragments"
)

head(da_peaks)

plot1 <- Seurat::VlnPlot(
  object = pbmc,
  features = rownames(da_peaks)[1],
  pt.size = 0.1,
  idents = c("CD4 Memory", "CD14 Mono")
)

plot2 <- Seurat::FeaturePlot(
  object = pbmc,
  features = rownames(da_peaks)[1],
  pt.size = 0.1
)

plot1 | plot2

fc <- Signac::FoldChange(pbmc, ident.1 = "CD4 Naive", ident.2 = "CD14 Mono")
head(fc)

open_cd4naive <- rownames(da_peaks[da_peaks$avg_log2FC > 0.5, ])
open_cd14mono <- rownames(da_peaks[da_peaks$avg_log2FC < -0.5, ])

closest_genes_cd4naive <- Signac::ClosestFeature(pbmc, regions = open_cd4naive)
closest_genes_cd14mono <- Signac::ClosestFeature(pbmc, regions = open_cd14mono)

head(closest_genes_cd4naive)
head(closest_genes_cd14mono)

############################
# PLOTTING GENOMIC REGIONS #
############################

# Set Plotting Order
levels(pbmc) <- c("CD4 Naive", "CD4 Memory", "CD8 Naive", "CD8 Effector", "DN T", 
                  "NK CD56bright", "NK CD56Dim", "pre-B", "pro-B", "pDC", "DC", "CD14 Mono",
                  "CD16 Mono")

Signac::CoveragePlot(
  object = pbmc,
  region = rownames(da_peaks)[1],
  extend.upstream = 40000,
  extend.downstream = 20000
)